addons/stock/models/stock_quant.py

```diff
@@ -478,6 +575,17 @@ def _onchange_serial_number(self):
             if message:
                 return {'warning': {'title': _('Warning'), 'message': message}}
 
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        if not self.location_id:
+            product = self.product_id
+            if product.tracking in ['lot', 'serial']:
+                previous_quants = self.env['stock.quant'].search([('product_id', '=', product.id)], limit=1, order='create_date desc')
+                if previous_quants:
+                    self.location_id = previous_quants.location_id
+            elif not self.user_has_groups('stock.group_stock_multi_warehouses'):
+                self.location_id = self.env['stock.warehouse'].search([], limit=1).in_type_id.default_location_dest_id
```

Comment:
IMO, you should take in account the company_id fields in this onchange (if set). It allows to avoid to take a warehouse from a other company and then later trigger error from the _check_company (same idea for the previous quant maybe ?).